scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
 
     #------------------------------
     # initialize dist_matrix for output
-    dist_matrix = np.zeros((len(indices), N), dtype=DTYPE)
-    dist_matrix.fill(np.inf)
-    dist_matrix[np.arange(len(indices)), indices] = 0
+    if min_only:
+        dist_matrix = np.full(N, np.inf, dtype=DTYPE)
+        dist_matrix[indices] = 0
+    else:
+        dist_matrix = np.zeros((len(indices), N), dtype=DTYPE)
+        dist_matrix.fill(np.inf)
+        dist_matrix[np.arange(len(indices)), indices] = 0
 
     #------------------------------
     # initialize predecessors for output
     if return_predecessors:
-        predecessor_matrix = np.empty((len(indices), N), dtype=ITYPE)
-        predecessor_matrix.fill(NULL_IDX)
+        if min_only:
+            predecessor_matrix = np.empty((N), dtype=ITYPE)
+            predecessor_matrix.fill(NULL_IDX)
+            source_matrix = np.empty((N), dtype=ITYPE)
```

Comment:
there may be some benefit to typing `source matrix`, but it is also true that the original Cython function wasn't thoroughly typed, so that job could perhaps fall to a future PR

Reply:
I'm not familiar enough with ctyping to know if the other reshape operations and things that are needed in this python function to return things in the appropriate python format are compatible with this.. i think we should leave this to another PR as you suggested.